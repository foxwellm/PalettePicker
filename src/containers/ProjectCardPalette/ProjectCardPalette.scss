.share-palette {
  margin: 10px 5px 0 10px;
  height: 20px;
  width: 20px;
  background: url('./images/share.png');
  background-size: cover;
  border: none;
  cursor: pointer;

  &:hover {
    background: url('./images/share-hover.png');
    background-size: cover;
  }
}

.delete-palette {
  margin: 10px 5px 0 5px;
  height: 20px;
  width: 20px;
  background: url('./images/trash.png');
  background-size: cover;
  border: none;
  cursor: pointer;

  &:hover {
    background: url('./images/trash-hover.png');
    background-size: cover;
  }
}

.name-text {
  width: 75px;
  text-align: left;
  margin: auto 10px auto 0;
}

.palette-color-div {
  margin: 5px 0;
}

.copied-text {
  margin-top: 13px;
  margin-bottom: 0;
  color: red;
}
