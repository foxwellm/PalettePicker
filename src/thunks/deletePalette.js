import { setLoading, setError, removePalette } from '../actions';
import { fetchData } from '../utils/api';

export const deletePalette = (id) => {
  return async (dispatch) => {
    try {
      dispatch(setLoading(true));
      await fetchData(`/api/v1/palettes/${id}`, 'DELETE');
      dispatch(removePalette(id));
    } catch (error) {
      dispatch(setError(error.message));
    }
    dispatch(setLoading(false));
  }
}