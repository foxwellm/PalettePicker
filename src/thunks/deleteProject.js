import { setLoading, setError, removeProject } from '../actions';
import { fetchData } from '../utils/api';

export const deleteProject = (id) => {
  return async (dispatch) => {
    try {
      dispatch(setLoading(true));
      await fetchData(`/api/v1/projects/${id}`, 'DELETE');
      dispatch(removeProject(id));
    } catch (error) {
      dispatch(setError(error.message));
    }
    dispatch(setLoading(false));
  }
}