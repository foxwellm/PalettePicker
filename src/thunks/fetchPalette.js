import { setLoading, setError, setPalettes } from '../actions';
import { fetchData } from '../utils/api';

export const fetchPalette = (id) => {
  return async (dispatch) => {
    try {
      dispatch(setLoading(true));
      const palette = await fetchData(`/api/v1/palettes/${id}`, 'GET');
      dispatch(setPalettes([palette]));
    } catch (error) {
      dispatch(setError(error.message));
    }
    dispatch(setLoading(false));
  };
};