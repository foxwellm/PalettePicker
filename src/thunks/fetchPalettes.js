import { setError } from '../actions';
import { fetchData } from '../utils/api';

export const fetchPalettes = (allProjects) => {
  return (dispatch) => {
    const promisedPalettes = allProjects.map(async palette => {
      try {
        let allPalettes = await fetchData(`/api/v1/projects/${palette.id}/palettes`, 'GET');
        allPalettes.sort((a, b) => {
          if (a.updated_at > b.updated_at) return -1;
          if (b.updated_at < a.updated_at) return 1;
          return 0;
        });
        return allPalettes
      } catch (error) {
        dispatch(setError(error.message));
      }
    })
    return Promise.all(promisedPalettes);
  }
}