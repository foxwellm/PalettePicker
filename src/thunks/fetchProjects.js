import { setLoading, setError, setProjects } from '../actions';
