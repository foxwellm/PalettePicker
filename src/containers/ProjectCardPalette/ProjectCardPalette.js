import React, { Component } from 'react';
import { deletePalette } from '../../thunks/deletePalette';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { CopyToClipboard } from 'react-copy-to-clipboard';

export class ProjectCardPalette extends Component {
  constructor(props) {
    super(props);
    this.state = {
      copied: false
    }
  }

  deleteCardPalette = async (e) => {
    await this.props.deletePalette(parseInt(e.target.id));
  }

  shareCardPalette = () => {
    this.setState({ copied: true })
    setTimeout(() => {
      this.setState({ copied: false })
    }, 2000);
  }

  render() {
    const { name, color1, color2, color3, color4, color5, id } = this.props;
    return (
      <div className='card-palette'>
        <h4 className='name-text'>{name}</h4>
        <div style={{ backgroundColor: color1 }} className='palette-color-div'></div>
        <div style={{ backgroundColor: color2 }} className='palette-color-div'></div>
        <div style={{ backgroundColor: color3 }} className='palette-color-div'></div>
        <div style={{ backgroundColor: color4 }} className='palette-color-div'></div>
        <div style={{ backgroundColor: color5 }} className='palette-color-div'></div>
        <CopyToClipboard text={window.location.href + `palette/${id}`}
          onCopy={this.shareCardPalette}>
          <button className='share-palette'></button>
        </CopyToClipboard>
        <button onClick={this.deleteCardPalette} id={id} className='delete-palette'></button>
        {this.state.copied && <p className='copied-text'>Copied</p>}
      </div>
    )
  }
}

export const mapDispatchToProps = (dispatch) => ({
  deletePalette: (id) => dispatch(deletePalette(id)),
});

ProjectCardPalette.propTypes = {
  deletePalette: PropTypes.func,
}

export default connect(null, mapDispatchToProps)(ProjectCardPalette);