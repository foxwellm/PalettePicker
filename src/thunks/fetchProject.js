import { setLoading, setError, setProjects, setPalettes } from '../actions';
import { fetchData } from '../utils/api';
import { fetchPalettes } from './fetchPalettes';

export const fetchProject = (id) => {
  return async (dispatch) => {
    try {
      dispatch(setLoading(true));
      const project = await fetchData(`/api/v1/projects/${id}`, 'GET');
      dispatch(setProjects([project]));
      const projectPalettes = await dispatch(fetchPalettes([project]));
      dispatch(setPalettes(projectPalettes));
    } catch (error) {
      dispatch(setError(error.message));
    }
    dispatch(setLoading(false));
  };
};