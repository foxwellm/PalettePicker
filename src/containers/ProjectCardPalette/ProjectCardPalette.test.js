import React from 'react';
import { ProjectCardPalette, mapDispatchToProps } from './ProjectCardPalette';
import { shallow } from 'enzyme';
import { deletePalette } from '../../thunks/deletePalette';
jest.mock('../../thunks/deletePalette');

describe('ProjectCardPalette', () => {
  let wrapper;
  let mockDeletePalette;
  let mockPalette;
  let mockEvent;

  beforeEach(() => {
    mockEvent = { target: { id: '1'} };
    mockDeletePalette = jest.fn();
    mockPalette = {
      id: 1,
      name: 'Palette 1',
      color1: '#ffffff',
      color2: '#ffffff',
      color3: '#ffffff',
      color4: '#ffffff',
      color5: '#ffffff'
    };

    wrapper = shallow(
      <ProjectCardPalette
        deletePalette={mockDeletePalette}
        id={mockPalette.id}
        name={mockPalette.name}
        color1={mockPalette.color1}
        color2={mockPalette.color2}
        color3={mockPalette.color3}
        color4={mockPalette.color4}
        color5={mockPalette.color5}
      />
    )
  });

  it('should match the correct snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it.skip('should call deletePalette on click when deleteCardPalette is called', async () => {
    await wrapper.find('button').simulate('click');
    expect(mockDeletePalette).toHaveBeenCalledWith(parseInt(mockEvent.target.id));
  });

  describe('mapDispatchToProps', () => {
    it('should call dispatch when deletePalette is called', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = deletePalette();
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.deletePalette();
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  });

});