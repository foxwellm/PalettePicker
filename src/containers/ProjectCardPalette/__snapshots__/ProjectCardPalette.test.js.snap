// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectCardPalette should match the correct snapshot 1`] = `
<div
  className="card-palette"
>
  <h4
    className="name-text"
  >
    Palette 1
  </h4>
  <div
    className="palette-color-div"
    style={
      Object {
        "backgroundColor": "#ffffff",
      }
    }
  />
  <div
    className="palette-color-div"
    style={
      Object {
        "backgroundColor": "#ffffff",
      }
    }
  />
  <div
    className="palette-color-div"
    style={
      Object {
        "backgroundColor": "#ffffff",
      }
    }
  />
  <div
    className="palette-color-div"
    style={
      Object {
        "backgroundColor": "#ffffff",
      }
    }
  />
  <div
    className="palette-color-div"
    style={
      Object {
        "backgroundColor": "#ffffff",
      }
    }
  />
  <CopyToClipboard
    onCopy={[Function]}
    text="http://localhost/palette/1"
  >
    <button
      className="share-palette"
    />
  </CopyToClipboard>
  <button
    className="delete-palette"
    id={1}
    onClick={[Function]}
  />
</div>
`;
